pkg/handler/iam_handler_test.go

```diff
@@ -0,0 +1,565 @@
+							Expression: fmt.Sprintf("request.time < timestamp('%s')", time.Now().Add(1*time.Hour).Format(time.RFC3339)),
+						},
+					},
+				},
+			},
+			wantProjectsPolicy: &iampb.Policy{},
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx := context.Background()
+
+			// Setup fake servers.
+			addrOrg, connOrg := testutil.FakeGRPCServer(t, func(s *grpc.Server) {
+				resourcemanagerpb.RegisterOrganizationsServer(s, tc.orgsServer)
+			})
+			defer connOrg.Close()
+			addrFol, connFol := testutil.FakeGRPCServer(t, func(s *grpc.Server) {
+				resourcemanagerpb.RegisterFoldersServer(s, tc.foldersServer)
+			})
+			defer connFol.Close()
+			addrPro, connPro := testutil.FakeGRPCServer(t, func(s *grpc.Server) {
+				resourcemanagerpb.RegisterProjectsServer(s, tc.projectsServer)
+			})
+			defer connPro.Close()
+
+			// Setup fake clients.
+			fakeOrgClient, err := resourcemanager.NewOrganizationsClient(ctx, option.WithGRPCConn(connOrg))
+			if err != nil {
+				t.Fatalf("creating client for fake at %q: %v", addrOrg, err)
+			}
+			fakeFolClient, err := resourcemanager.NewFoldersClient(ctx, option.WithGRPCConn(connFol))
+			if err != nil {
+				t.Fatalf("creating client for fake at %q: %v", addrFol, err)
+			}
+			fakeProClient, err := resourcemanager.NewProjectsClient(ctx, option.WithGRPCConn(connPro))
+			if err != nil {
+				t.Fatalf("creating client for fake at %q: %v", addrPro, err)
+			}
+
+			h, err := NewIAMHandler(
+				ctx,
+				fakeOrgClient,
+				fakeFolClient,
+				fakeProClient,
+				WithRetry(retry.WithMaxRetries(0, retry.NewFibonacci(500*time.Millisecond))),
+			)
+			if err != nil {
+				t.Fatalf("failed to create IAMHandler: %v", err)
+			}
+
+			// Run test.
+			gotPolicies, gotErr := h.Do(ctx, tc.request)
+			if diff := testutil.DiffErrString(gotErr, tc.wantErrSubstr); diff != "" {
+				t.Errorf("Process(%+v) got unexpected error substring: %v", tc.name, diff)
+			}
+			// Verify that the Policies are modified accordingly.
+			if diff := cmp.Diff(tc.wantPolicies, gotPolicies, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got diff (-want, +got): %v", tc.name, diff)
+			}
+
+			if diff := cmp.Diff(tc.wantOrgsPolicy, tc.orgsServer.policy, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got org policy diff (-want, +got): %v", tc.name, diff)
+			}
+
+			if diff := cmp.Diff(tc.wantFoldersPolicy, tc.foldersServer.policy, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got folder policy diff (-want, +got): %v", tc.name, diff)
+			}
+
+			if diff := cmp.Diff(tc.wantProjectsPolicy, tc.projectsServer.policy, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got project policy diff (-want, +got): %v", tc.name, diff)
+			}
+		})
+	}
+}
+
+type fakeOrgsServer struct {
+	resourcemanagerpb.UnimplementedOrganizationsServer
+
+	policy          *iampb.Policy
+	getIAMPolicyErr error
+	setIAMPolicyErr error
```

Comment:
To slightly reduce the duplication here, you could put these fields in a separate struct. And that separate struct will implement two func: `getIAMPolicy` and `setIAMPolicy`. By using this struct in all the fakes, you can reduce the duplicated logic.

Reply:
When registered the grpc conn, it got confused with which set/getIAMPolicy to use, so still need to define the set/getIAMPolicy for the server struct, have to figure out some other ways if any...

Reply:
Tried to put all unimplemented org/folder/projects server in the same struct, since there is a common TestIamPermissions API in all three servers, and it got confused which one to use. To avoid that we have to define TestIamPermissions, but I don't think it is a good idea.

Reply:
I think after you use `IAMClient` in your handler, we don't need separated fakes. We can just need to pick one, say `fakeProjectServer`, it will match the interface for all resources.

Reply:
Discussed offline.